### ISSUE TITLE
Time based rotation

### ISSUE THREAD

**Issue by arossert** [Preston-Landers/concurrent-log-handler#23]
As far as I can tell there is no way to configure the rotation by time like TimedRotatingFileHandler, will it be hard to add another handler that will rotate like TimedRotatingFileHandler but with the locking mechanism?

**Comment by Preston-Landers** [comment]
That's not available at the present time, but I don't see why it couldn't be added. I'll take a look when I get a chance. Pull requests are welcome too.

**Comment by arossert** [comment]
@Preston-Landers Thanks for the reply.
I was thinking of using the built-in `TimedRotatingFileHandler` handler for the time rotation handling and your `ConcurrentRotatingFileHandler` for the file locking capabilities.
Do you have any idea how we can do it easily? I'm willing to work on that and provide PR.

**Comment by jh11200** [comment]
不知道你们是不是真的外国人，总是用英语聊天。
我自己基于TimedRotatingFileHandler做了一些修改，请大佬指教下可不可行：
`
    
    import os
    import time
    from logging.handlers import TimedRotatingFileHandler
    
    
    class HxTimedRotatingFileHandler(TimedRotatingFileHandler):
    
        @property
        def dfn(self):
            currentTime = int(time.time())
            # get the time that this sequence started at and make it a TimeTuple
            dstNow = time.localtime(currentTime)[-1]
            t = self.rolloverAt - self.interval
            if self.utc:
                timeTuple = time.gmtime(t)
            else:
                timeTuple = time.localtime(t)
                dstThen = timeTuple[-1]
                if dstNow != dstThen:
                    if dstNow:
                        addend = 3600
                    else:
                        addend = -3600
                    timeTuple = time.localtime(t + addend)
            dfn = self.rotation_filename(self.baseFilename + "." + time.strftime(self.suffix, timeTuple))
    
            return dfn
    
        def shouldRollover(self, record):
            """
            是否应该执行日志滚动操作：
            1、存档文件已存在时，执行滚动操作
            2、当前时间 >= 滚动时间点时，执行滚动操作
            """
            dfn = self.dfn
            t = int(time.time())
            if t >= self.rolloverAt or os.path.exists(dfn):
                return 1
            return 0
    
        def doRollover(self):
            """
            执行滚动操作
            1、文件句柄更新
            2、存在文件处理
            3、备份数处理
            4、下次滚动时间点更新
            """
            if self.stream:
                self.stream.close()
                self.stream = None
            # get the time that this sequence started at and make it a TimeTuple
    
            dfn = self.dfn
    
            # 存档log 已存在处理
            if not os.path.exists(dfn):
                self.rotate(self.baseFilename, dfn)
    
            # 备份数控制
            if self.backupCount > 0:
                for s in self.getFilesToDelete():
                    os.remove(s)
    
            # 延迟处理
            if not self.delay:
                self.stream = self._open()
    
            # 更新滚动时间点
            currentTime = int(time.time())
            newRolloverAt = self.computeRollover(currentTime)
            while newRolloverAt <= currentTime:
                newRolloverAt = newRolloverAt + self.interval
    
            # If DST changes and midnight or weekly rollover, adjust for this.
            if (self.when == 'MIDNIGHT' or self.when.startswith('W')) and not self.utc:
                dstAtRollover = time.localtime(newRolloverAt)[-1]
                dstNow = time.localtime(currentTime)[-1]
                if dstNow != dstAtRollover:
                    if not dstNow:  # DST kicks in before next rollover, so we need to deduct an hour
                        addend = -3600
                    else:           # DST bows out before next rollover, so we need to add an hour
                        addend = 3600
                    newRolloverAt += addend
            self.rolloverAt = newRolloverAt

`

**Comment by arossert** [comment]
@jh11200 this will not use the ConcurrentRotatingFileHandler locking mechanism.

**Comment by jh11200** [comment]
> @jh11200 this will not use the ConcurrentRotatingFileHandler locking mechanism.

是没有用到，对于日志，我只要求写入最新的那个文件就行，写入的顺序无所谓。不加锁不知道会不会有什么影响。目前我已经放线上环境运行了。在观察有么有问题

**Comment by jh11200** [comment]
> @jh11200 this will not use the ConcurrentRotatingFileHandler locking mechanism.

我观察了两天，我是以when=midnight进行分割的，没什么问题。日志可以正常写入一个文件里面。

**Comment by arossert** [comment]
Did you test it with multiple processes writing to the same file?

What is the difference between this and the original time base logger?

**Comment by jh11200** [comment]
> Did you test it with multiple processes writing to the same file?
> 
> What is the difference between this and the original time base logger?

我这里是用uwsgi启动了3个进程，线上跑了这么多天，没发现问题。
区别在于：
1、在`def shouldRollover:` 增加了判断：如果应该归档的文件已存在时也执行一次归档操作。
2、在`def doRollover:`修改了判断：只在归档文件不存在时才执行创建操作。

**Comment by xujiang1** [comment]
@jh11200 我在flask中使用了你的HxTimedRotatingFileHandler 调用为HxTimedRotatingFileHandler (logfile, "midnight", 1, 7) 由于flask在重载代码的机制下为双进程，依然会出现
`PermissionError: [WinError 32] 另一个程序正在使用此文件，进程无法访问。: 'D:\\logdev\\normal.log' -> 'D:\\logdev\\normal.log.2019-12-05'`
 这基本是多个进程占用文件导致文件无法重命名，我不知道你是如何解决的，望告知

**Comment by rootVIII** [comment]
Is this going to be added? I was just about to open an issue to ask for a feature request... I need to roll logs daily instead of based on the size. Thanks

**Comment by ugvibib** [comment]
> Is this going to be added? I was just about to open an issue to ask for a feature request... I need to roll logs daily instead of based on the size. Thanks

https://github.com/ugvibib/pyloghandlers
I modify it simply. you see see🤪

**Comment by tshirtman** [comment]
Actually went on and experimented with that, quickly tested with an existing project needing the feature, and it seems to behave as expected, the script that many subprocesses all using the same logger config, and tailing the resulting log file now displays the same result as the console output.

**Comment by Preston-Landers** [comment]
Hi folks, 

Good news for anyone who is interested in this feature. I've written a new implementation of a timed rotating handler that should address the problems I mentioned in PR#48. It's a direct subclass of the base `TimedRotatingFileHandler` and composes an instance of `ConcurrentRotatingFileHandler` to reuse locking and other behaviors. It also coordinates the rollover time with other processes. It supports both time-based and size-based rotation (defaults to time-only, every 1 hour, like the base class).

More details can be found in the updated README.md on the branch `timed_rotating_v2`. You can review PR#55 and leave comments there.

I would be happy if anyone could test this out and provide any feedback. There is already some unit test coverage for the new features but additional testing and validation is always great. If no problems surface within a week or so I may go ahead and publish this as a new version. Please also note that as of this new version we're requiring Python 3.6 or higher.

Use a line in `requirements.txt` like this to install the branch:
```
git+https://github.com/Preston-Landers/concurrent-log-handler.git@timed_rotating_v2#egg=concurrent_log_handler
```

Or you can grab a wheel from here:
https://github.com/Preston-Landers/concurrent-log-handler/releases/tag/0.9.23-alpha-1

Note that the original PR for this feature has been rebased to the newest code and is still available on a branch `concurrent_timed_rotating_file_handler` if you want to test or compare that as well.

**Comment by Preston-Landers** [comment]
Merged and will be released as 0.9.23. Thanks!
